job_controller/pod.go

```diff
@@ -253,3 +271,187 @@ func (jc *JobController) GetPodSlices(pods []*v1.Pod, replicas int, logger *log.
+	if !ok {
+		return fmt.Errorf("job is not a runtime.Object type")
+	}
+	jobKey, err := KeyFunc(metaObject)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for job object %#v: %v", job, err))
+		return err
+	}
+	expectationPodsKey := GenExpectationPodsKey(jobKey, rt)
+	err = jc.Expectations.ExpectCreations(expectationPodsKey, 1)
+	if err != nil {
+		return err
+	}
+	logger := commonutil.LoggerForReplica(metaObject, rt)
+
+	// Set type and index for the worker.
+	labels := jc.GenLabels(metaObject.GetName())
+	labels[jc.Controller.GetReplicaTypeLabelKey()] = rt
+	labels[jc.Controller.GetReplicaIndexLabelKey()] = index
+
+	if masterRole {
+		labels[LabelJobRole] = "master"
+	}
+
+	podTemplate := spec.Template.DeepCopy()
+
+	// Set name for the template.
+	podTemplate.Name = GenGeneralName(metaObject.GetName(), rt, index)
+
+	if podTemplate.Labels == nil {
+		podTemplate.Labels = make(map[string]string)
+	}
+
+	for key, value := range labels {
+		podTemplate.Labels[key] = value
+	}
+
+	if err := jc.Controller.SetClusterSpec(job, podTemplate, rt, index); err != nil {
+		return err
+	}
+
+	// Submit a warning event if the user specifies restart policy for
+	// the pod template. We recommend to set it from the replica level.
+	if podTemplate.Spec.RestartPolicy != v1.RestartPolicy("") {
+		errMsg := "Restart policy in pod template will be overwritten by restart policy in replica spec"
+		logger.Warning(errMsg)
+		jc.Recorder.Event(runtimeObject, v1.EventTypeWarning, podTemplateRestartPolicyReason, errMsg)
+	}
+	setRestartPolicy(podTemplate, spec)
+
+	// if gang-scheduling is enabled:
+	// 1. if user has specified other scheduler, we report a warning without overriding any fields.
+	// 2. if no SchedulerName is set for pods, then we set the SchedulerName to "kube-batch".
+	if jc.Config.EnableGangScheduling {
+		if isNonGangSchedulerSet(replicas) {
+			errMsg := "Another scheduler is specified when gang-scheduling is enabled and it will not be overwritten"
+			logger.Warning(errMsg)
+			jc.Recorder.Event(runtimeObject, v1.EventTypeWarning, podTemplateSchedulerNameReason, errMsg)
+		} else {
+			podTemplate.Spec.SchedulerName = gangSchedulerName
+		}
+	}
+
+	err = jc.Controller.CreatePod(job, podTemplate)
+	if err != nil && errors.IsTimeout(err) {
+		// Pod is created but its initialization has timed out.
+		// If the initialization is successful eventually, the
+		// controller will observe the creation via the informer.
+		// If the initialization fails, or if the pod keeps
+		// uninitialized for a long time, the informer will not
+		// receive any update, and the controller will create a new
+		// pod when the expectation expires.
+		return nil
+	} else if err != nil {
+		return err
+	}
+	return nil
+}
+
+func setRestartPolicy(podTemplateSpec *v1.PodTemplateSpec, spec *common.ReplicaSpec) {
+	if spec.RestartPolicy == common.RestartPolicyExitCode {
+		podTemplateSpec.Spec.RestartPolicy = v1.RestartPolicyNever
```

Comment:
Add a comment explaining the reason for this (restartPolicyExitCode is not supported in podTemplateSpec).